File: nvpl_blas/f77/ssymm.f
!******************************************************************************
! Content:
!     This example demonstrates use of API as below:
!     SSYMM 
!
!******************************************************************************
      program   SSYMM_MAIN
*
      implicit none

      include "nvpl_blas_f77_blas.fi"

      character*1     side, uplo
      parameter       (side='L', uplo='U')
      integer         m, n, i, j
      integer         lda, ldb, ldc
      parameter       (m=2, n=2,
     &                 lda=2, ldb=2, ldc=2)
      integer         asize, bsize, csize
      parameter       (asize=m*m, bsize=m*n, csize=m*n)
      real            alpha, beta
      parameter       (alpha=1.0, beta=1.0)
      real            a(lda,m), b(ldb,n), c(ldc,n)

*       External Subroutines
      external        print_smatrix, fill_smatrix
*
*      Executable Statements
*
      print*
      print 99
*
*       Print input data
      print*
      print 100, m, n, lda, ldb, ldc, side, uplo
      print 101, alpha, beta
      call fill_smatrix(a,m,m,lda,uplo,'N')
      call fill_smatrix(b,m,n,ldb,'F','N')
      call fill_smatrix(c,m,n,ldc,'F','N')
      call print_smatrix(a,m,m,lda,'A')
      call print_smatrix(b,m,n,ldb,'B')
      call print_smatrix(c,m,n,ldc,'C')
*
*      Call SSYMM subroutine
      call SSYMM(side,uplo,m,n,alpha,a,lda,b,ldb,beta,c,ldc)
*
      print*
      print 102, 'After SSYMM operation:'
      call print_smatrix(c,m,n,ldc,'C')

      stop
 99   format('Example: SSYMM for symmetric matrix-matrix'
     &       ' multiplication')
 100  format('#### args: m=',i1,', n=',i1,', lda=',i1,
     &       ', ldb=',i1,', ldc=',i1,', side=',a1,', uplo=',a1)
 101  format(11x,'alpha=',f4.1,'  beta=',f4.1)
 102  format(A)     
      end

File: nvpl_blas/c/ssymm.c
/******************************************************************************
 * Content: 
 *     This example demonstrates use of API as below:
 *     cblas_ssymm
 *
 ******************************************************************************/
#include "example_helper.h"

int main() {
    nvpl_int_t M = 2;
    nvpl_int_t N = 2;
    nvpl_int_t lda = 2;
    nvpl_int_t ldb = 2;
    nvpl_int_t ldc = 2;
    float alpha = 1.0f;
    float beta = 1.0f;
    enum CBLAS_ORDER order = CblasRowMajor;
    enum CBLAS_SIDE side = CblasLeft;
    enum CBLAS_UPLO uplo = CblasUpper;
    float * A = NULL;
    float * B = NULL;
    float * C = NULL;

    printf("\nExample: cblas_ssymm for symmetric matrix-matrix multiplication\n\n");
    printf("#### args: m=%" PRId64 ", n=%" PRId64 ", lda=%" PRId64 ", ldb=%" PRId64 ", ldc=%" PRId64 ", "
            "order=%c, side=%c, uplo=%c\n", (int64_t)M, (int64_t)N, (int64_t)lda, (int64_t)ldb,
            (int64_t)ldc, order_to_char(order), side_to_char(side), uplo_to_char(uplo));
    printf("           alpha=%g, beta=%g\n", alpha, beta);

    nvpl_int_t rowsA = (side == CblasLeft) ? M : N;
    nvpl_int_t colsA = rowsA; 
    nvpl_int_t rowsB = M;
    nvpl_int_t colsB = N;
    nvpl_int_t rowsC = M;
    nvpl_int_t colsC = N;

    // allocate memory
    A = (float *)malloc(lda * colsA  * sizeof(float));
    B = (float *)malloc(ldb * colsB  * sizeof(float));
    C = (float *)malloc(ldc * colsC  * sizeof(float));

    // fill data
    fill_smatrix(A, rowsA, colsA, lda, order, UPLO_TO_FILL_MODE(uplo), CblasNonUnit);
    fill_smatrix(B, rowsB, colsB, ldb, order, Full, CblasNonUnit);
    fill_smatrix(C, rowsC, colsC, ldc, order, Full, CblasNonUnit);

    // print input data
    print_smatrix(A, rowsA, colsA, lda, "A", order);
    print_smatrix(B, rowsB, colsB, ldb, "B", order);
    print_smatrix(C, rowsC, colsC, ldc, "C", order);

    // call cblas_ssymm
    cblas_ssymm(order, side, uplo, M, N, alpha, A, lda, B, ldb, beta, C, ldc);

    // print result
    printf("\nAfter cblas_ssymm operation:\n");
    print_smatrix(C, rowsC, colsC, ldc, "C", order);

    // release memory
    free(A);
    free(B);
    free(C);
    return 0;
}
